bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -741,6 +796,46 @@ private static boolean checkIsMapType(BType sourceType, BMapType targetType, Lis
         }
     }
 
+    private static boolean checkIsMapType(Object sourceVal, BType sourceType, BMapType targetType,
+                                          List<TypePair> unresolvedTypes) {
+        BType targetConstrainedType = targetType.getConstrainedType();
+        switch (sourceType.getTag()) {
+            case TypeTags.MAP_TAG:
+                return checkConstraints(((BMapType) sourceType).getConstrainedType(), targetConstrainedType,
+                                        unresolvedTypes);
+            case TypeTags.RECORD_TYPE_TAG:
+                MapValue sourceMapValue = (MapValue) sourceVal;
+                BRecordType recType = (BRecordType) sourceType;
+
+                for (BField field : recType.getFields().values()) {
+                    if (!Flags.isFlagOn(field.flags, Flags.READONLY)) {
+                        if (!checkIsType(field.type, targetConstrainedType, unresolvedTypes)) {
+                            return false;
+                        }
+                        continue;
+                    }
+
+                    BString name = StringUtils.fromString(field.name);
+
+                    if (Flags.isFlagOn(field.flags, Flags.OPTIONAL) && !sourceMapValue.containsKey(name)) {
+                        continue;
+                    }
+
+                    if (!checkIsLikeType(sourceMapValue.get(name), targetConstrainedType)) {
```

Comment:
What's the reason to call `checkIslikeType()` for readonly fields and `checkIsType()` for non-readonly fields?

Reply:
A `readonly` field cannot be updated, and whatever the value that was initially set is guaranteed to be immutable. 

For an immutable value, is type is the same as is like type (as though the "inherent type" changed). 

So if there is a `readonly` field, we know that it can never change (the reference or the value). So if the value specified for that field looks like the target type it would mean that it belongs to the target type.

This is similar to how we call check-is-like-type when check-is-type is used with immutable values.